### DISCUSSION TITLE
Can someone try compiling bootloader and firmware .bin for 401CC?

### DISCUSSION THREAD

**Discussion by Ganesharyal24894** [blackmagic-debug/blackmagic#1907] [Q&A]
So I am trying to find the problem as to why my STM32F401CC is not able to be converted as BMP(discussed here issues #1899 ). Initially, I was thinking there could be some problem with the hardware, so I ordered another board and received it today, but even that Blackpill board does not work. So now I am sure that it is not a hardware problem. Can someone try to compile the bootloader and firmware `.bin` for Blackpill (401CC) and send it here? I just want to verify that it is a problem with compilation on my system.

**Comment by ALTracer** [comment]
Compiled from current main, v1.10.0-1031-g3a3558d7, and flashed to WeActStudio MiniSTM32F411CE `blackpill-f411ce` with 25MHz HSE crystal, confirmed booting/enumerating/answering to GDB and BMDA. All targets enabled, RTT enabled. No diagnostic logging or other fancy stuff.

OEM ST bootloader:
`$ meson setup build-f401cc_maskrom --cross-file cross-file/blackpill-f401cc.ini -Dbmd_bootloader=false`
[build-f401cc_maskrom.zip](https://github.com/user-attachments/files/16715153/build-f401cc_maskrom.zip)
`$ dfu-util -d 0483:df11 -S 375F334C3032 --alt 0 -s 0x08000000:leave -D build-f401cc_maskrom/blackmagic_blackpill_f401cc_firmware.bin`

BMD bootloader:
`$ meson setup build-f401cc --cross-file cross-file/blackpill-f401cc.ini`
[build-f401cc.zip](https://github.com/user-attachments/files/16715175/build-f401cc.zip)
`$ dfu-util -d 0483:df11 -S 375F334C3032 --alt 0 -s 0x08000000:leave -D build-f401cc/blackmagic_blackpill_f401cc_bootloader.bin`
`$ dfu-util -d 1d50:6017 -S 375F334C3032 --alt 0 -s 0x08004000:leave -D build-f401cc/blackmagic_blackpill_f401cc_firmware.bin`

In case of troubles, use ELF files for inception debugging your blackpills via an stlink.

**Reply by Ganesharyal24894** [reply]
Okay so I tried flashing this and I get the same result which is "Invalid device descriptor" in device manager so now I am pretty sure that the compilation is happening correctly on my system too. Now my doubt is, this is some Windows BS and I don't know what is inception debugging that you are suggesting
BTW you said you tested this on a **411ce**, is the firmware same for both **411ce** and **401cc**  or is it just a typo?

**Reply by ALTracer** [reply]
I tested this on f411ce because that's all I've got. I have never seen any f401cc boards, and they are long deprecated by WeActStudios themselves anyways.
The firmware is almost the same to the point where f401cc firmware runs on f411ce hardware, so I was able to verify my builds for you. I usually compile and flash for f411ce because it has a higher clock (84->96 MHz), more SRAM (96->128 KiB) and more internal Flash (256->512 KiB), *and I need that for faster operations and diagnostic logging and other developer-level instrumentation.*
If you want help, then provide visibility. Post photo of board wired and crashed with chip markings visible, post screenshot of flasher tool with chip ID visible (0x423, 0x433, 0x431) and verification success.
If you don't want help, then use the stlink to debug the (working!) Blackmagic Firmware on your board, collect the crash stack backtrace and analyze it. Then bring a 100MHz oscilloscope and look at USB D+/D- wires and decode the failing enumeration.

**Reply by dragonmux** [reply]
Only thing we can add to ALTracer's response is: Check the markings on the crystal/oscillator; the firmware needs to be using a matching PLL configuration to the clock source fitted external to the processor. If it's set for a 8MHz source but is getting a 25MHz, for example, the processor will effectively be overclocked and that can cause this sort of failure. Checking what is fitted to check against what the firmware is expecting is a cheap and fast thing to eliminate one of the possible issues here.

**Reply by Ganesharyal24894** [reply]
I don't know about where you live but for my location 401cc is easier and cheaper to get, I don't think they are deprecated though and I really wish I had a 100Mhz Oscilloscope XD.
It was indeed some problem with Windows, just checked it on a UP squared board (linux mint) and it works perfectly fine. You guys are really helpful.
Thank you.

**Reply by dragonmux** [reply]
WeAct Studio, the people behind the Black Pill boards, discontinued both of the F401 models (see [the Black Pill repository](https://github.com/WeActStudio/WeActStudio.MiniSTM32F4x1)). This would mean that your board must be be a clone as they don't make or sell those variants any more. This is also where them being deprecated comes from

**Reply by Ganesharyal24894** [reply]
I don't know it is either a clone or the seller is selling aged inventory but at least the ST chip is original with 25MHz Crystal and it has WeAct studio marking but with version 3.0, as long as it works it is fine for me.

**Reply by sidprice** [reply]
This sounds like the same problem I am seeing on Windows with the Blackpill F411. It seems the request to read the Device Descriptor is failing when the BP boots the BMD firmware. I am going to hook up my USN analyzer and see what it reveals.

**Reply by Ganesharyal24894** [reply]
> This sounds like the same problem I am seeing on Windows with the Blackpill F411. It seems the request to read the Device Descriptor is failing when the BP boots the BMD firmware. I am going to hook up my USN analyzer and see what it reveals.

Are you able to see your board in DFU mode when you upload the bootloader only?

**Reply by sidprice** [reply]
@Ganesharyal24894 Yes, I have the Blackpill being enumerated in bootloader mode (ST bootloader) and the firmware is flashing using that. When the firmware boots it fails to enumerate as a BMP. @dragonmux and I are investigating the issue.

Note: I am using the ST Bootloader at this time.

As I said, my USB analyzer clearly shows the Device Descriptor being send by the Blackpill running the BMD firmware is invalid.

**Reply by Ganesharyal24894** [reply]
I think this is probably happening because of some recent changes made in the repository, because I did this just two or three weeks ago, and it was working perfectly fine with Windows as well.

**Reply by dragonmux** [reply]
Having worked with Sid, we determined that it was down to a change in USB reset behaviour introduced to support STM32H7's in the DWC2 driver. Guarding that new behaviour to not compile for F4's solved the problem and restored functionality of the firmware. Please try current `main` - you will need to recompile both the firmware and bootloader and reload both on your board. When you pull `main`, you will need to update locm3 too with `git submodule update deps/libopencm3` once you have pulled the latest `main` for BMD.

Apologies for all the false turns and uncertainty with not being able to tell if this was a firmware bug, a your hardware bug, or a Windows bug till now.

**Reply by Ganesharyal24894** [reply]
Yep, It is visible now. Thank you to all of you. Still I have a question though how was it working with Linux ?

**Reply by dragonmux** [reply]
We don't think it was.. we strongly suspect this has been broken since #1892 as the bug in question causes the firmware to miss-drive the USB peripheral.

**Reply by Ganesharyal24894** [reply]
No, I am 100% sure it was, since I and @ALTracer both had checked it. He had checked and then sent me the compiled firmware in this same discussion.

**Reply by dragonmux** [reply]
The bug in question only triggers if a USB Reset occurs under certain conditions - which are possible under Linux, but just not guaranteed to occur. This is why we say we suspect this has just been broken since that PR.

**Reply by Ganesharyal24894** [reply]
Got it now. Once again, really Thank you.

**Reply by ALTracer** [reply]
Since you've discovered and fixed a bug of extraneous USBOTG_FS reset in F4 drivers, should I delete my bug-exhibiting tested builds which worked on Linux for me but break on Windows for @Ganesharyal24894 ? Or upload new ones, keeping these for reproducibility with a real USB analyzer? Or CI builds and nightly links are enough? 
To be clear, I was seeing a new kernel warning then, `Device not responding to setup address.`, and then it enumerated the second time successfully. Only when booting normally through BMD bootloader, but not when leaving from MaskROM into BMD bootloader or booting with PA0 pressed down. It's gone now.
I would prefer the new developments to be its own accepted answer, but whatever.

**Reply by Ganesharyal24894** [reply]
@ALTracer I have unmarked that as an answer, just in case anyone comes across that.

**Reply by dragonmux** [reply]
We'll close the discussion as resolved for now in that you've been able to prove it was not your compilation environment, but rather a bug that's now been fixed in libopencm3. It can be re-opened later if necessary.
